libs/langgraph/langgraph/prebuilt/tool_node.py

```diff
@@ -127,6 +144,37 @@ def _validate_tool_call(self, call: ToolCall) -> Optional[ToolMessage]:
         else:
             return None
 
+    def _inject_state(
+        self, tool_call: ToolCall, input: Union[list[AnyMessage], dict[str, Any]]
+    ) -> ToolCall:
+        state_args = _get_state_args(self.tools_by_name[tool_call["name"]])
+        if state_args and not isinstance(input, dict):
+            required_fields = list(state_args.values())
+            if (
+                len(required_fields) == 1
+                and required_fields[0] == "messages"
+                or required_fields[0] is None
+            ):
+                input = {"messages": input}
+            else:
+                err_msg = (
+                    f"Invalid input to ToolNode. Tool {tool_call['name']} requires "
+                    f"graph state dict as input."
+                )
+                if any(state_field for state_field in state_args.values()):
+                    required_fields_str = ", ".join(f for f in required_fields if f)
+                    err_msg += f" State should contain fields {required_fields_str}."
+                raise ValueError(err_msg)
+        tool_call_copy: ToolCall = copy(tool_call)
```

Comment:
Instead of copy and mutate should we do:

```
tool_call_copy = {
  ***tool_call,
  "args": {
    ...
    }
}
```

Reply:
whats the diff? or do you mean stylistically

Reply:
nvm